### PR TITLE
Import vCard from Attachment and from Native

### PR DESCRIPTION
**Pull request by murilopereirame**
When user receives a vCard as attachment an import option is displayed inside the attachment bubble, allowing the user to preview and import the users contained inside that vCard.

The app handles both vCard mime types, text/vcard and text/x-vcard

close [#6469](https://github.com/tutao/tutanota/issues/6469)

### CODE REVIEW

File: src/mail/view/MailViewer.ts [hunk 1]
```diff
@@ -727,6 +729,25 @@ export class MailViewer implements Component<MailViewerAttrs> {
 		}
 		return false
 	}
+
+	private async handleAttachmentImport(file: TutanotaFile) {
```
Comment (charlag) [inline src/mail/view/MailViewer.ts:733]: from a first glance all of this seems to be mostly unrelated to `MailViewer`. I wonder if we could do  something like
`Mailviewer -> MailViewerViewModel.importContactFromFile() -> ContactImporter.importContactFromFile()` and do all of this there?

File: src/mail/view/MailViewerViewModel.ts [hunk 2]
```diff
@@ -1043,6 +1042,19 @@ export class MailViewerViewModel {
...
+	async downloadAndParseVCard(file: TutanotaFile) {
+		const { vCardFileToVCards } = await import("../../contacts/VCardImporter.js")
+		file = (await this.cryptoFacade.enforceSessionKeyUpdateIfNeeded(this._mail, [file]))[0]
+		try {
+			const fileData = await this.fileController.getFileData(file, file.mimeType ?? VCARD_MIME_TYPES.X_VCARD)
+			const decoder = new TextDecoder("utf-8")
+
+			return vCardFileToVCards(decoder.decode(fileData))
+		} catch (e) {
+			await Dialog.message("errorDuringFileOpen_msg")
```
Comment (charlag) [inline src/mail/view/MailViewerViewModel.ts:1054]: dialogs don't belong here, we can throw a `UserError` though or return some kind of result

File: src/native/main/WebCommonNativeFacade.ts [hunk 3]
```diff
@@ -37,32 +41,54 @@ export class WebCommonNativeFacade implements CommonNativeFacade {
...
-					recipients,
-					subject || (files.length > 0 ? files[0].name : ""),
-					signatureModule.appendEmailSignature(text || "", logins.getUserController().props),
-					files,
-					undefined,
-					undefined,
-					true, // we want emails created in this method to always default to saving changes
-				)
-			}
+				const allFilesAreVCards = files.filter((file) => Object.values<string>(VCARD_MIME_TYPES).includes(file.mimeType)).length === files.length
```
Comment (charlag) [inline src/native/main/WebCommonNativeFacade.ts:48]: consider making a helper "isVcardFile` or `isVcardMimeType`, it will make it much more readable

File: src/native/main/WebCommonNativeFacade.ts [hunk 4]
```diff
@@ -37,32 +41,54 @@ export class WebCommonNativeFacade implements CommonNativeFacade {
...
+
+				let willImport = false
+				if (allFilesAreVCards) {
+					willImport = await Dialog.choice(
+						() => "We detected some contact files, do you want to import or attach them?",
```
Comment (charlag) [inline src/native/main/WebCommonNativeFacade.ts:53]: we should clarify the text and also not forget to translate it

File: src/native/main/WebCommonNativeFacade.ts [hunk 5]
```diff
@@ -152,4 +178,51 @@ export class WebCommonNativeFacade implements CommonNativeFacade {
...
+				const dataFile = await fileApp.readDataFile(file.location)
+				if (dataFile == null) continue
+
+				const blob = new Blob([dataFile.data], { type: file.mimeType })
+				const buffer = await blob.arrayBuffer()
+				const decoder = new TextDecoder("utf-8")
+				const parsedVCards = vCardFileToVCards(decoder.decode(buffer))
+				if (parsedVCards == null) continue
+
+				const contactMembership = assertNotNull(logins.getUserController().user.memberships.find((m) => m.groupType === GroupType.Contact))
```
Comment (charlag) [inline src/native/main/WebCommonNativeFacade.ts:200]: We've made a helper on our branch, maybe consider using copying it too?
https://github.com/tutao/tutanota/blob/0a89bb1ff36690d8896deba3544dce64e361c658/src/api/main/UserController.ts#L186-L188

File: test/tests/api/worker/facades/ContactFacadeTest.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import o from "@tutao/otest"
+import { ContactFacade } from "../../../../../src/api/worker/facades/lazy/ContactFacade.js"
+import td, { object, verify } from "testdouble"
+import { EntityClient } from "../../../../../src/api/common/EntityClient.js"
+import { downcast } from "@tutao/tutanota-utils"
+import { createFilledContact } from "../../../contacts/VCardExporterTest.js"
+
+o.spec("ContactFacadeTest", function () {
+	const entityClient: EntityClient = object()
```
Comment (charlag) [inline test/tests/api/worker/facades/ContactFacadeTest.ts:9]: you want to assign it in `before()` so that it's reset every time

File: test/tests/api/worker/facades/ContactFacadeTest.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+	o.before(() => {
+		entityClient.setupMultipleEntities = downcast(td.function())
```
Comment (charlag) [inline test/tests/api/worker/facades/ContactFacadeTest.ts:12]: you don't need to do that I think

File: src/gui/base/Dialog.ts [hunk 8]
```diff
@@ -1006,6 +1007,30 @@ export class Dialog implements ModalComponent {
 		return dialog
 	}
 
+	/**
+	 * Show a dialog with a preview of a given list of contacts
+	 * @param contacts The contact list to be previewed
+	 * @param okAction The action to be executed when the user press the import button
+	 */
+	static importVCardDialog(contacts: Contact[], okAction: (dialog: Dialog) => unknown): Dialog {
```
Comment (charlag) [inline src/gui/base/Dialog.ts:1015]: I'm sorry but this does not belong here. Dialog class if for all kinds of generic dialogs, this is a very very specific one.

File: src/gui/base/Dialog.ts [hunk 9]
```diff
@@ -1006,6 +1007,30 @@ export class Dialog implements ModalComponent {
...
+		return Dialog.showActionDialog({
+			title: () => lang.get("importVCard_action"),
+			child: () =>
+				m(
+					".mt-s",
+					contacts.map((contact) => {
```
Comment (charlag) [inline src/gui/base/Dialog.ts:1021]: Since we are going to use it from native import as well I was wondering if it would be a better idea to use a real `List` instead. We just need to give it a static state, we don't need a `ListModel`

Comment (murilopereirame) [reply src/gui/base/Dialog.ts]: Implemented

File: src/gui/base/Dialog.ts (hunk 8 above)
Comment (murilopereirame) [reply src/gui/base/Dialog.ts]: I see this Dialog as generic as the ```showRequestPasswordDialog```, but I moved it to implement the List.

File: src/native/main/WebCommonNativeFacade.ts (hunk 5 above)
Comment (murilopereirame) [reply src/native/main/WebCommonNativeFacade.ts]: Sure

File: src/native/main/WebCommonNativeFacade.ts (hunk 4 above)
Comment (murilopereirame) [reply src/native/main/WebCommonNativeFacade.ts]: Added the keys

File: src/native/main/WebCommonNativeFacade.ts (hunk 3 above)
Comment (murilopereirame) [reply src/native/main/WebCommonNativeFacade.ts]: Used a helper function to get the attachment type and compare with the enum. Is more readable now imo

File: src/mail/view/MailViewer.ts (hunk 1 above)
Comment (murilopereirame) [reply src/mail/view/MailViewer.ts:735]: Part of it can be moved to another file, but downloading & parsing the file needs to be done here as we enforce the sessionkey update to the file based on the email.

File: src/mail/view/MailViewer.ts [hunk 10]
```diff
@@ -727,6 +729,25 @@ export class MailViewer implements Component<MailViewerAttrs> {
...
+		const contactListId = await this.viewModel.contactModel.getContactListId()
+
+		if (vCardData == null || contactListId == null) {
+			return Dialog.message("importVCardError_msg")
+		}
+
+		const contactMembership = assertNotNull(locator.logins.getUserController().user.memberships.find((m) => m.groupType === GroupType.Contact))
+		const contacts = vCardListToContacts(vCardData, contactMembership.group)
+
+		return Dialog.importVCardDialog(contacts, async (dialog) => {
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/mail/view/MailViewer.ts]: removed async from function's signature

File: src/gui/AttachmentBubble.ts [hunk 11]
```diff
@@ -115,6 +117,13 @@ export class AttachmentDetailsPopup implements ModalComponent {
 				help: "remove_action",
 			})
 		}
+		if (attrs.file_import) {
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/gui/AttachmentBubble.ts]: I added because we have shortcuts to all actions, maybe some people will use it?

File: src/gui/AttachmentBubble.ts [hunk 12]
```diff
@@ -22,6 +22,8 @@ export type AttachmentBubbleAttrs = {
 	download: Thunk | null
 	open: Thunk | null
 	remove: Thunk | null
+	file_import: Thunk | null
+	isVCard: boolean
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/gui/AttachmentBubble.ts]: For sure, an Enum will probably be better

File: src/gui/AttachmentBubble.ts [hunk 13]
```diff
@@ -22,6 +22,8 @@ export type AttachmentBubbleAttrs = {
 	download: Thunk | null
 	open: Thunk | null
 	remove: Thunk | null
+	file_import: Thunk | null
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/gui/AttachmentBubble.ts]: Changed to camelCase

File: src/file/FileController.ts [hunk 14]
```diff
@@ -113,6 +119,13 @@ export abstract class FileController {
 		await this.observeProgress(this.doDownload([file], DownloadPostProcessing.Open))
 	}
 
+	async getFileData(file: TutanotaFile, mime: string): Promise<ArrayBuffer> {
+		const dataFile = await this.getAsDataFile(file)
+		const blob = new Blob([dataFile.data], { type: mime })
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/file/FileController.ts]: Thanks for the hint, the function isn't necessary anymore. I need to decode the bytes and I can do it from datafile.data.

File: src/api/worker/WorkerLocator.ts [hunk 15]
```diff
@@ -393,6 +397,10 @@ export async function initLocator(worker: WorkerImpl, browserData: BrowserData)
 		const { ConfigurationDatabase } = await import("./facades/lazy/ConfigurationDatabase.js")
 		return new ConfigurationDatabase(locator.user)
 	})
+	locator.contactFacade = lazyMemoized(async () => {
+		const { ContactFacade } = await import("./facades/lazy/ContactFacade.js")
+		return new ContactFacade(new EntityClient(entityRestClient))
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/api/worker/WorkerLocator.ts:402]: Here I wasn't sure which one to use. I saw that we don't use the entity rest with cache on Calendar stuff, why we don't use and why use on contacts?

File: app-android/app/src/main/java/de/tutao/tutanota/AndroidFileFacade.kt [hunk 16]
```diff
@@ -331,7 +331,19 @@ class AndroidFileFacade(
 
 	@Throws(IOException::class)
 	override suspend fun readDataFile(filePath: String): DataFile? {
-		throw Error("FileFacade.readDataFile should not be used in Android")
+		val uri = Uri.parse(filePath)
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply app-android/app/src/main/java/de/tutao/tutanota/AndroidFileFacade.kt:334]: Sorry, didn't get that. Do you mean inside of our folders (encrypted/decrypted)? I think that it's fine to read arbitrary files since the user is giving us the path, I don't believe that we'll have problems to read from directories outside /temp

[the remark this answers is not in the document]
Comment (charlag) [reply app-android/app/src/main/java/de/tutao/tutanota/AndroidFileFacade.kt:334]: We should restrict it to out encrypted/decrypted temp directories, yes. We can try to allow content provider URLs too but I think we anyway copy the files to our temp directories when we receive them, do we not?

File: src/api/worker/WorkerLocator.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (charlag) [reply src/api/worker/WorkerLocator.ts:402]: You mean CalendarFacade? We do have both caching and non-caching one there to be able to bypass the cache and get some things before we process entity updates. As a rule of a thumb you should use the caching one.

File: src/gui/AttachmentBubble.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (charlag) [reply src/gui/AttachmentBubble.ts]: It's probably fine, it's just if you don't receive contact files every day you won't memorize it

File: src/gui/base/Dialog.ts (hunk 8 above)
Comment (charlag) [reply src/gui/base/Dialog.ts]: Yeah that one should also not be here to be honest :sweat_smile:

File: src/contacts/ContactImporter.ts [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+					} satisfies ListAttrs<Contact, ImportContactRowHolder>),
+				),
+			),
+		okAction,
+		allowCancel: true,
+		okActionTextId: "import_action",
+	})
+}
+
+class ImportContactVirtualRow implements VirtualRow<Contact> {
```
Comment (charlag) [inline src/contacts/ContactImporter.ts:103]: I know you are following an existing pattern but in this case we can collapse both row and viewHolder into viewHolder

File: src/mail/view/MailViewer.ts (hunk 1 above)
Comment (charlag) [reply src/mail/view/MailViewer.ts:735]: I don't think parsing needs to be done here, only downloading/enforcing session key stuff and then we can pass it on to importer

File: src/native/main/WebCommonNativeFacade.ts [hunk 18]
```diff
@@ -45,20 +43,17 @@ export class WebCommonNativeFacade implements CommonNativeFacade {
 				editor.show()
 			} else {
 				const files = await fileApp.getFilesMetaData(filesUris)
-				const allFilesAreVCards = files.filter((file) => Object.values<string>(VCARD_MIME_TYPES).includes(file.mimeType)).length === files.length
+				const allFilesAreVCards = files.filter((file) => getAttachmentType(file.mimeType) === AttachmentType.CONTACT).length === files.length
```
Comment (charlag) [inline src/native/main/WebCommonNativeFacade.ts:46]: `filter` -> `every`

**Review by charlag — APPROVED** [review]
